doc/rfc008-managing-apalache.md

```diff
@@ -0,0 +1,343 @@
+
+Regardless of the Quint release, we expect the latest version of Apalache to be
+installed. We could use the Github REST API to check the latest Apalache release
+and abort if the installed version differs.
+
+Pros:
+
+- Easy management, no need to pin specific versions from Quint.
+
+Cons:
+
+- Prone to break if users have an outdated version of Quint.
+  - We could address this by also checking the latest release of Quint, and
+    alerting users to upgrade.
+- Prone to break if users downgrade their installation of Quint.
+
+#### 3.3.3 Pin a specific Apalache version
+
+Each Quint release pins a specific version of Apalache.
+
+Pros:
+
+- Enables user to downgrade Quint, while maintaining compatibility with Apalache.
+
+Cons:
+
+- Users don't receive Apalache bugfix releases, unless we cut another Quint
+  release that pins the updated dependency.
+
+#### 3.3.4 Pin a minor release branch
+
+Each Quint release pins an Apalache minor release, but allows patch-level
+updates. This would correspond to a [tilde range](https://docs.npmjs.com/cli/v6/using-npm/semver#tilde-ranges-123-12-1)
+`~major.minor.patch` in common package managers.
+
+Pros:
+
+- Users receive patch versions of Apalache while maintaining compatibility.
+- Enables user to downgrade Quint, while maintaining compatibility with Apalache.
+
+Cons:
+
+- There may be an Apalache minor version bump that does not break compatibility
+  Quint. We would need to cut another Quint release that pins the updated minor
+  release.
+
+## 4. Proposed Solution
+
+We choose managed options for obtaining and launching Apalache from Quint, while
+still providing a non-managed option:
+
+1. Primarily, Quint uses a Quint-managed, on-demand instance of Shai:
+   1. If not present, Quint fetches a compatible Apalache release from Github,
+      using [Github's REST API endpoint][] and/or [`octokit/request.js`][] and
+      unpacks it into a local installation directory.
+      - This can be scripted etirely from NodeJS, foregoing the need to wrap a
+        third-party package manager binary.
+   2. Apalache's JRE dependency is taken care of by adding a check for a `java`
+      executable in `$PATH` to [the `apalache-mc` runner script](https://github.com/informalsystems/apalache/blob/df7adb8b42b6487de9764162f338935121d07a3c/src/universal/bin/apalache-mc#L53).
+      - This shall print instructions for obtaining a JRE, if none is detected.
+2. Quint launches an on-demand instance of this local installation by spawning
+   `apalache-mc server` in a separate process[^1].
+   - Such a process is spawned for each invocation of the `verify` command, and
+     torn down immediately after the RPC response.
+     - This seems to be a reasonable tradeoff given Apalache's memory
+       consumption and JVM startup time.
+     - We can later change this to a long-lived server, if we use a more
+       stateful approach of interacting with Shai (e.g., via the [transition
+       explorer API][]).
```

Comment:
An option not discussed here afaict is tying apalache's lifetime to the quint processes lifetime, e.g., by:

- Having an interactive terminal session for checking specs, akin to `sbt`
- Managing the server from the vs-code extension.

I think these can be postponed as possibly later feature, but the should be noted here for consideration.

Reply:
I really meant this to apply regardless of how verification is invoked. Could be the CLI, VSCode, or another frontend. I think discussing front-end options is orthogonal to this document, since all of them need to somehow spawn and connect to a Shai instance, and ideally this ADR should provide a transparent solution for all of them.

This also came up in @bugarela's review, I will revise this paragraph to be less ambiguous.